### DISCUSSION TITLE
Unclear on join needed with CTE

### DISCUSSION THREAD

**Discussion by pdc1** [sqlalchemy/sqlalchemy#10623] [Usage Questions] [answered]
I am trying to combine two queries on a book database. One for author+title, returning date read, and one for the most recent date the same author has been read (for any book).

I thought it might be more efficient to use a CTE to get all the authors and then do the two subqueries, but it's getting complex and I might be kidding myself that this is gaining anything, so that could be a valid answer (i.e. just do two queries).

A snippet from the database definition:
```python
class BookDB(MappedAsDataclass, Base):
    __tablename__ = "book"
    id: Mapped[int_pk] = mapped_column(init=False)
    author: Mapped[str50]
    title: Mapped[str150]
    date_read: Mapped[Optional[date]]
```

What I have for the query:
```python
sq = (
    select(BookDB.author, BookDB.date_read, BookDB.title)
    .where(BookDB.author == "<author>")
    .cte()
)
sq2 = (
    select(sq.c.author, func.max(sq.c.date_read))
    .subquery()
)
q = (
    select(sq.c.author, sq.c.date_read, sq.c.title, sq2)
    .where(sq.c.title == "<title>")
)
with Session(engine) as session:
    row = session.execute(q).one_or_none()
    if row:
        print(">", row)
```

While this appears to work, I get a warning:
`SAWarning: SELECT statement has a cartesian product between FROM element(s) "anon_2" and FROM element "anon_1".  Apply join condition(s) between each element to resolve.`

I then tried adding `select_from()` and `join()` but the only thing I found that worked is:
```python
q = (
    select(sq.c.author, sq.c.date_read, sq.c.title, sq2)
    .select_from(sq)
    .join(sq2), sq.c.author == sq2.c.author)
    .where(sq.c.title == "<title>")
)
```

And that seems unnecessary to me since `sq` already consists entirely of the same author.

Any suggestions? Is there a better way to handle this? 

Thanks!

**Comment by CaselIT** [comment, marked as answer]
hi,

Well you need to join the two subquery, otherwise you get a cartesian product, like the warning says. 
In this case it's not an issue since there is only one row returned in `sq2`. Sqlalchemy does not know this.

You can use `join(sql2, sa.true()` to make a no-op join while removing the warning. But I don't think doing your second query will create issues for the db.
